src/operation.rs

```diff
@@ -838,64 +852,77 @@ fn process_binary_operation(
             }
         }
         BinaryOp::Eq() => {
-            /// Take an iterator of pairs of RichTerm, the common environments of all left
-            /// components of these pairs and all right components, the final environment,
-            /// and build a Term which evaluates to `Bool(true)` if and only if all the pairs are
-            /// equals
-            fn eq_all<T>(
-                it: T,
-                env1: &Environment,
-                env2: &Environment,
-                env: &mut Environment,
-            ) -> Term
-            where
-                T: Iterator<Item = (RichTerm, RichTerm)>,
-            {
-                let subeqs: Vec<RichTerm> = it
-                    .map(|(t1, t2)| {
-                        let t1_var = t1.closurize(env, env1.clone());
-                        let t2_var = t2.closurize(env, env2.clone());
-                        Term::Op2(BinaryOp::Eq(), t1_var, t2_var).into()
-                    })
-                    .collect();
-                // lists.all (fun x => x) subeqs
-                Term::App(
-                    mk_app!(stdlib::lists::all(), mk_term::id()),
-                    Term::List(subeqs).into(),
-                )
-            }
+            let mut env = Environment::new();
+
+            let c1 = Closure {
+                body: RichTerm {
+                    term: t1,
+                    pos: pos1,
+                },
+                env: env1,
+            };
+            let c2 = Closure {
+                body: RichTerm {
+                    term: t2,
+                    pos: pos2,
+                },
+                env: env2,
+            };
 
-            let mut env: Environment = HashMap::new();
-            let res = match (*t1, *t2) {
-                (Term::Bool(b1), Term::Bool(b2)) => Term::Bool(b1 == b2),
-                (Term::Num(n1), Term::Num(n2)) => Term::Bool(n1 == n2),
-                (Term::Str(s1), Term::Str(s2)) => Term::Bool(s1 == s2),
-                (Term::Lbl(l1), Term::Lbl(l2)) => Term::Bool(l1 == l2),
-                (Term::Sym(s1), Term::Sym(s2)) => Term::Bool(s1 == s2),
-                (Term::Record(m1), Term::Record(m2)) => {
-                    let (left, center, right) = merge::hashmap::split(m1, m2);
+            match eq(&mut env, c1, c2) {
+                EqResult::Bool(b) => Ok(Closure::atomic_closure(Term::Bool(b).into())),
+                EqResult::Eqs(fst, snd, eqs) => Ok(Closure {
+                    body: RichTerm::new(Term::Op2(BinaryOp::MultiEq(eqs), fst, snd), pos_op),
+                    env,
+                }),
+            }
+        }
+        BinaryOp::MultiEq(eqs) => {
```

Comment:
I'm wondering, though. Instead of making an operation which stores a bunch of values, how about making a binary operation which returns a function:

- It forces the two values on the stack, and test them for equality
  - If not equal, pop the top of the stack, push `false` (that's the function `const false`)
  - If equal, then evaluate the value on the top of the stack (that's `id`)

Then
- The standard equality test `a == b` is `%eqThen% a b True`
- You can push stuff onto the stack itself to store computation (some care has to be taken, as you can only use the top of the stack, but the point is that we are not using an extraneous storage that the virtual machine doesn't control)

An alternative may be, instead of returning a boolean function, to take a continuation as the last argument

While we are fiddling with this semantics, it's worth considering.

Reply:
Currently `MultiEq` is a binary operator (it is applied to the two next operands to force and test for equality in the sequence of equalities) which stores a continuation of the remaining equalities to test as an inner parameter `Vec<(RichTerm, RichTerm)>`. What about the following changes:

- instead of storing `eqs` inside the `MultiEq` operator, push the whole list as one element on the stack (or as a sequence of elements using a stack marker to determine the end). This is the continuation of equalities to test.
- when processing the next equality, instead of returning an application `MultiEq t1 t2` which would go back to the main eval loop, push the continuation `(MultiEq, t2)` on the stack, and goes on to evaluate `t1`, we can inline this step in `MultiEq` and directly push the continuation `(MultiEq, t2)` on the stack and go on evaluating `t1`, so that we don't create the intermediate `MultiEq t1 t2`.

Then `MultiEq` would test for equality its two evaluated argument, and proceed as before. If new equalities needs to be added, we can insert them in the `eqs` continuation on the top of the stack. If one equality is false, we discard the top of the stack (or everything up to the marker) and return false. If one equality is true, we pop the next one to compute from the top of the stack, or return true if there's no more.

This would be more or less the same process as the current implementation, but storing `eqs` and next terms to proceed on the stack rather than as a `MultiEq` application, which is what your suggestion is about, if I understand correctly. Would that do ?

Reply:
You don't need the `MultiEq` to be an n-ary operator, because it's annoying to write. The binary-operator-returning-a-function idea avoids this: you just have a binary operator, and the third element does not need to be forced (well, I mean, you need to eventually enter it if the first equality is true, since it's basically a thunk to the boolean value that you need).

On the other hand, this approach does build up quite a bit of heap values, which may not be very nice. If you want to use purely the stack, you need either a stack marker or a count (it is worth noting that pushing everything to the stack, with a bit of care, flattens all the equalities to test, so you really have a single loop going on, which is nice). What I would do, there, is that `MultiEq ; t1 :: t2 :: stack`, if `t1 == t2` would step to `MultiEq ; stack`. Still a binary operation.

You have to be careful to clear the stack when `t1 /= t2`, though.

Reply:
> On the other hand, this approach does build up quite a bit of heap values, which may not be very nice. If you want to use purely the stack, you need either a stack marker or a count (it is worth noting that pushing everything to the stack, with a bit of care, flattens all the equalities to test, so you really have a single loop going on, which is nice). What I would do, there, is that `MultiEq ; t1 :: t2 :: stack`, if `t1 == t2` would step to `MultiEq ; stack`. Still a binary operation.
> 
> You have to be careful to clear the stack when `t1 /= t2`, though.

Yes, this is what I was referring to, even if I strangely wanted to store the continuation as one big vector on the stack first, rather than directly as stack consecutive elements, I don't know why. Anyway, this is implemented in c82eed5.